### ISSUE TITLE
Using estimators created by `tf.keras.estimator.model_to_estimator` in `tf.estimator.train_and_evaluate` causes a memory leak of sorts

### ISSUE THREAD

**Issue by zmjjmz** [tensorflow/tensorflow#19407]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes, see provided gist
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Debian Jessie
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: ('v1.8.0-0-g93bc2e2072', '1.8.0')
- **Python version**: 2.7.9
- **Bazel version (if compiling from source)**: n/a
- **GCC/Compiler version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a
- **Exact command to reproduce**: Run provided gist

### Describe the problem
When using keras models with `train_and_evaluate`, you need to convert them into estimators first. The best way to do this is `tf.keras.estimator.model_to_estimator`. However, when `train_and_evaluate` is called with those estimators, the default graphs that are created end up referenced in `_GRAPH_LEARNING_PHASES` and never deallocated (so sort of a memory leak). This is due to a call to `keras.backend.set_learning_phase` that's done as part of the created `model_fn`, which uses the default graph as a key into `_GRAPH_LEARNING_PHASES`. Since that graph changes every so often in `train_and_evaluate`, that key changes so there's a new key inserted without the old ones being removed.

Or at leas, I'm 80% certain that's what's going on. You can see in the provided code a very simple Keras model turned into an estimator and then run with a hook that prints out memory usage and the size of `_GRAPH_LEARNING_PHASES`. It increases over time!

While the leak is small in the provided example, it has caused me great strife in real usage.
### Source code / logs

The source code and a sample run are provided in [this gist](https://gist.github.com/zmjjmz/392ead713f19db025390f6d8de17bde2). It should be easy to run the python script yourself however you'll need to have `numpy` and `psutil` installed.

**Comment by Dref360** [comment]
I found that adding : 

```python
class MemoryLogHook(tf.train.SessionRunHook):
    def begin(self):
        K._GRAPH_LEARNING_PHASES = {}
        K._GRAPH_UID_DICTS = {}
```

fixes the issue. (will stabilize around 1.35e+9) We probably need to clear those before calling the model_fn.

**Comment by zmjjmz** [comment]
Hey thanks! This workaround works for me, although I imagine it's probably not the best solution :)

**Comment by bhack** [comment]
Also I think that this workaround will not work with `mode_to_estimator` cause in that case the model_fn it is not explicitly built.

**Comment by facaiy** [comment]
ping @fchollet

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 19 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 34 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 49 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 64 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sdll** [comment]
Any updates on this?

**Comment by kumariko** [comment]
@zmjjmz We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.6 which is latest stable version of TF and let us know if the issue still persists in newer versions. Thanks!
